modal/functions.py

```diff
@@ -362,21 +387,33 @@ async def _map_invocation(
     have_all_inputs = False
     num_inputs = 0
     num_outputs = 0
+
+    def count_update():
+        if count_update_callback is not None:
+            count_update_callback(num_outputs, num_inputs)
+
     pending_outputs: Dict[str, int] = {}  # Map input_id -> next expected gen_index value
     completed_outputs: Set[str] = set()  # Set of input_ids whose outputs are complete (expecting no more values)
 
     input_queue: asyncio.Queue = asyncio.Queue()
 
-    async def create_input(arg: Any) -> api_pb2.FunctionPutInputsItem:
+    async def create_input(argskwargs):
         nonlocal num_inputs
         idx = num_inputs
         num_inputs += 1
-        item = await _create_input(arg, kwargs, client, idx=idx)
-        return item
+        (args, kwargs) = argskwargs
+        return await _create_input(args, kwargs, client, idx)
+
+    async def input_iter():
+        while 1:
+            raw_input = await raw_input_queue.get()
+            if raw_input is None:  # end of input sentinel
+                return
+            yield raw_input  # args, kwargs
 
     async def drain_input_generator():
         # Parallelize uploading blobs
-        proto_input_stream = stream.iterate(input_stream) | pipe.map(
+        proto_input_stream = stream.iterate(input_iter()) | pipe.map(
```

Comment:
Wouldn't be too hard I think, and might actually let us refactor this code. Will take a look in a followup. Also thinking about moving all of the mapping code to a separate module since this module is getting clogged with too much functionality imo